src/components/swap/PendingModalContent/Logos.tsx

```diff
@@ -0,0 +1,119 @@
+import { Currency } from '@uniswap/sdk-core'
+import { LoaderV3 } from 'components/Icons/LoadingSpinner'
+import CurrencyLogo from 'components/Logo/CurrencyLogo'
+import styled, { useTheme } from 'styled-components/macro'
+
+import { ReactComponent as PapersIcon } from '../../../assets/svg/papers-text.svg'
+
+export const LogoContainer = styled.div`
+  height: 48px;
+  width: 48px;
+  position: relative;
+  display: flex;
+  border-radius: 50%;
+  overflow: visible;
+`
+
+const CurrencyLoaderContainer = styled.div<{ $visible: boolean; asBadge: boolean }>`
+  z-index: 2;
+  border-radius: 50%;
+  transition: all ${({ theme }) => `${theme.transition.duration.medium} ${theme.transition.timing.inOut}`};
+  position: absolute;
+  opacity: ${({ $visible }) => ($visible ? 1 : 0)};
+  height: ${({ asBadge }) => (asBadge ? '20px' : '100%')};
+  width: ${({ asBadge }) => (asBadge ? '20px' : '100%')};
+  bottom: ${({ asBadge }) => (asBadge ? '-4px' : 0)};
+  right: ${({ asBadge }) => (asBadge ? '-4px' : 0)};
+  outline: ${({ theme, asBadge }) => (asBadge ? `2px solid ${theme.background}` : '')};
+`
+
+const RaisedCurrencyLogo = styled(CurrencyLogo)`
+  z-index: 1;
+`
+
+export function CurrencyLoader({
+  visible,
+  currency,
+  asBadge = false,
+}: {
+  visible: boolean
+  currency: Currency | undefined
+  asBadge?: boolean
+}) {
+  return (
+    <CurrencyLoaderContainer
+      asBadge={asBadge}
+      $visible={visible}
+      data-testid={`pending-modal-currency-logo-${currency?.symbol}`}
+    >
+      <RaisedCurrencyLogo currency={currency} size="100%" />
+    </CurrencyLoaderContainer>
+  )
+}
+
+const PinkCircle = styled.div<{ $visible: boolean }>`
+  position: absolute;
+  display: flex;
+  height: 100%;
+  width: 100%;
+  border-radius: 50%;
+  align-items: center;
+  justify-content: center;
+  background-color: ${({ theme }) => theme.userThemeColor};
+  z-index: 1;
+  opacity: ${({ $visible }) => ($visible ? 1 : 0)};
+  transition: opacity ${({ theme }) => `${theme.transition.duration.medium} ${theme.transition.timing.inOut}`};
```

Comment:
so there are a bunch of components in here that have the same property of opacity 1 <-> 0 with a transition timing. Thoughts on making a general `FadeIn` component that is a simple div wrapper with these two properties?

Reply:
im wondering also if this component should handle mounting/unmounting children when they're done with the exit animation and no longer visible. you could probably make use of `useUnmountingAnimation` to handle that logic - that way we wouldn't have to actually mount all these elements on the page if they're not actually rendered. you could then amend your top-level logic to be like:

```
{currentStep === ConfirmModalState.APPROVING_TOKEN && <PaperIcon />}
```